File: python-side/atm/algorithms/hamming.py
"""Codigo de Hamming generico con paridad par y correccion de un bit."""

from dataclasses import dataclass


@dataclass(frozen=True)
class DecodeResult:
    data_bits: str
    corrected: bool
    syndrome: int
    valid: bool
    error: str | None = None


def _validate_bits(bits: str) -> None:
    if any(bit not in "01" for bit in bits):
        raise ValueError("la cadena solo puede contener bits 0 y 1")


def required_parity_bits(message_bits: int) -> int:
    """Retorna el menor r tal que m + r + 1 <= 2**r."""
    if message_bits < 0:
        raise ValueError("message_bits no puede ser negativo")
    r = 0
    while message_bits + r + 1 > (1 << r):
        r += 1
    return r


def encode(data_bits: str) -> str:
    """Inserta bits de paridad en posiciones 1, 2, 4, 8, ..."""
    _validate_bits(data_bits)
    if not data_bits:
        return ""

    parity_count = required_parity_bits(len(data_bits))
    codeword = [0] * (len(data_bits) + parity_count + 1)  # indice cero no usado
    data_index = 0

    for position in range(1, len(codeword)):
        if position & (position - 1):
            codeword[position] = int(data_bits[data_index])
            data_index += 1

    for parity_position in (1 << i for i in range(parity_count)):
        parity = 0
        for position in range(1, len(codeword)):
            if position & parity_position:
                parity ^= codeword[position]
        codeword[parity_position] = parity

    return "".join(str(bit) for bit in codeword[1:])


def _syndrome(codeword: list[int], parity_count: int) -> int:
    syndrome = 0
    for parity_position in (1 << i for i in range(parity_count)):
        parity = 0
        for position in range(1, len(codeword)):
            if position & parity_position:
                parity ^= codeword[position]
        if parity:
            syndrome += parity_position
    return syndrome


def decode(encoded_bits: str, message_length: int) -> DecodeResult:
    """Verifica, corrige un error y extrae exactamente message_length bits."""
    _validate_bits(encoded_bits)
    if message_length < 0:
        raise ValueError("message_length no puede ser negativo")

    parity_count = required_parity_bits(message_length)
    expected_length = message_length + parity_count
    if len(encoded_bits) != expected_length:
        return DecodeResult(
            "", False, 0, False,
            f"longitud Hamming invalida: se esperaban {expected_length} bits",
        )
    if not encoded_bits:
        return DecodeResult("", False, 0, True)

    codeword = [0] + [int(bit) for bit in encoded_bits]
    syndrome = _syndrome(codeword, parity_count)
    corrected = False

    if syndrome:
        if syndrome >= len(codeword):
            return DecodeResult(
                "", False, syndrome, False,
                "el sindrome apunta fuera de la trama; error no corregible",
            )
        codeword[syndrome] ^= 1
        corrected = True
        if _syndrome(codeword, parity_count):
            return DecodeResult(
                "", False, syndrome, False,
                "la trama conserva paridad invalida despues de corregir",
            )

    data = "".join(
        str(codeword[position])
        for position in range(1, len(codeword))
        if position & (position - 1)
    )
    return DecodeResult(data[:message_length], corrected, syndrome, True)

File: proyecto/go-side/internal/algorithms/hamming.go
// Package algorithms contiene los esquemas de integridad de la capa de Enlace.
package algorithms

import (
	"fmt"
	"strings"
)

// HammingResult describe la verificacion y posible correccion de una trama.
type HammingResult struct {
	DataBits  string
	Corrected bool
	Syndrome  int
	Valid     bool
	Error     string
}

func validateBits(bits string) error {
	for _, bit := range bits {
		if bit != '0' && bit != '1' {
			return fmt.Errorf("la cadena solo puede contener bits 0 y 1")
		}
	}
	return nil
}

// RequiredParityBits retorna el menor r tal que m+r+1 <= 2^r.
func RequiredParityBits(messageBits int) (int, error) {
	if messageBits < 0 {
		return 0, fmt.Errorf("messageBits no puede ser negativo")
	}
	r := 0
	for messageBits+r+1 > (1 << r) {
		r++
	}
	return r, nil
}

// HammingEncode inserta paridad par en las posiciones potencia de dos.
func HammingEncode(dataBits string) (string, error) {
	if err := validateBits(dataBits); err != nil {
		return "", err
	}
	if dataBits == "" {
		return "", nil
	}
	r, _ := RequiredParityBits(len(dataBits))
	codeword := make([]byte, len(dataBits)+r+1)
	dataIndex := 0
	// Las potencias de dos se reservan para paridad; el resto recibe los datos.
	for position := 1; position < len(codeword); position++ {
		if position&(position-1) != 0 {
			codeword[position] = dataBits[dataIndex] - '0'
			dataIndex++
		}
	}
	// Una paridad cubre los indices que tienen activo su bit correspondiente.
	for parityPosition := 1; parityPosition < (1 << r); parityPosition <<= 1 {
		parity := byte(0)
		for position := 1; position < len(codeword); position++ {
			if position&parityPosition != 0 {
				parity ^= codeword[position]
			}
		}
		codeword[parityPosition] = parity
	}
	var builder strings.Builder
	builder.Grow(len(codeword) - 1)
	for _, bit := range codeword[1:] {
		builder.WriteByte(bit + '0')
	}
	return builder.String(), nil
}

func hammingSyndrome(codeword []byte, parityCount int) int {
	// La suma de paridades fallidas forma la posicion binaria del error.
	syndrome := 0
	for parityPosition := 1; parityPosition < (1 << parityCount); parityPosition <<= 1 {
		parity := byte(0)
		for position := 1; position < len(codeword); position++ {
			if position&parityPosition != 0 {
				parity ^= codeword[position]
			}
		}
		if parity != 0 {
			syndrome += parityPosition
		}
	}
	return syndrome
}

// HammingDecode verifica, corrige un error y extrae messageLength bits.
func HammingDecode(encodedBits string, messageLength int) HammingResult {
	if err := validateBits(encodedBits); err != nil {
		return HammingResult{Valid: false, Error: err.Error()}
	}
	r, err := RequiredParityBits(messageLength)
	if err != nil {
		return HammingResult{Valid: false, Error: err.Error()}
	}
	expectedLength := messageLength + r
	if len(encodedBits) != expectedLength {
		return HammingResult{Valid: false, Error: fmt.Sprintf("longitud Hamming invalida: se esperaban %d bits", expectedLength)}
	}
	if encodedBits == "" {
		return HammingResult{Valid: true}
	}
	codeword := make([]byte, len(encodedBits)+1)
	for index := range encodedBits {
		codeword[index+1] = encodedBits[index] - '0'
	}
	syndrome := hammingSyndrome(codeword, r)
	corrected := false
	if syndrome != 0 {
		// Hamming SEC corrige un bit volteando la posicion indicada por el sindrome.
		if syndrome >= len(codeword) {
			return HammingResult{Valid: false, Syndrome: syndrome, Error: "el sindrome apunta fuera de la trama; error no corregible"}
		}
		codeword[syndrome] ^= 1
		corrected = true
		if hammingSyndrome(codeword, r) != 0 {
			return HammingResult{Valid: false, Syndrome: syndrome, Error: "la trama conserva paridad invalida despues de corregir"}
		}
	}
	var data strings.Builder
	data.Grow(messageLength)
	for position := 1; position < len(codeword); position++ {
		if position&(position-1) != 0 {
			data.WriteByte(codeword[position] + '0')
		}
	}
	value := data.String()
	return HammingResult{DataBits: value[:messageLength], Corrected: corrected, Syndrome: syndrome, Valid: true}
}
